src/instructions/segmentation.rs

```diff
@@ -1,210 +1,278 @@
+
+            unsafe fn set_reg(sel: SegmentSelector) {
+                #[cfg(feature = "inline_asm")]
+                asm!(concat!("mov ", $name, ", {0:x}"), in(reg) sel.0, options(nostack, preserves_flags));
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_ds(sel.0);
+                #[cfg(not(feature = "inline_asm"))]
+                crate::asm::$asm_load(sel.0);
+            }
+        }
+    };
 }
 
-/// Reload es segment register.
-///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid extra segment descriptor.
-#[inline]
-pub unsafe fn load_es(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov es, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+macro_rules! segment64_impl {
+    ($type:ty, $name:literal, $base:ty, $asm_rd:ident, $asm_wr:ident) => {
+        impl Segment64 for $type {
+            const BASE: Msr = <$base>::MSR;
+            unsafe fn read_base() -> VirtAddr {
+                #[cfg(feature = "inline_asm")]
+                {
+                    let val: u64;
+                    asm!(concat!("rd", $name, "base {}"), out(reg) val, options(nomem, nostack, preserves_flags));
+                    VirtAddr::new_unsafe(val)
+                }
+                #[cfg(not(feature = "inline_asm"))]
+                VirtAddr::new_unsafe(crate::asm::$asm_rd())
+            }
+
+            unsafe fn write_base(base: VirtAddr) {
+                #[cfg(feature = "inline_asm")]
+                asm!(concat!("wr", $name, "base {}"), in(reg) base.as_u64(), options(nostack, preserves_flags));
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_es(sel.0);
+                #[cfg(not(feature = "inline_asm"))]
+                crate::asm::$asm_wr(base.as_u64());
+            }
+        }
+    };
 }
 
-/// Reload fs segment register.
+/// Code Segment
 ///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid fs segment descriptor.
-#[inline]
-pub unsafe fn load_fs(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov fs, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+/// The segment base and limit are unused in 64-bit mode. Only the L (long), D
+/// (default operation size), and DPL (descriptor privilege-level) fields of the
+/// descriptor are recognized. So chaning the segment register can be used to
+/// change privilege level or enable/disable long mode.
+#[derive(Debug)]
+pub struct CS;
+impl Segment for CS {
+    get_reg_impl!("cs", x86_64_asm_get_cs);
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_fs(sel.0);
+    unsafe fn set_reg(sel: SegmentSelector) {
+        // Note this is special since we cannot directly move to cs. Instead we
+        // push the new segment selector and return value on the stack and use
+        // retfq to reload cs and continue at 1:.
```

Comment:
Perhaps we should make this a doc comment so that it shows up in the API docs too. It might be good to know that this behaves differently than reloading the other segment registers.

```suggestion
    /// Note this is special since we cannot directly move to cs. Instead we
    /// push the new segment selector and return value on the stack and use
    /// retfq to reload cs and continue at 1:.
    unsafe fn set_reg(sel: SegmentSelector) {
```

Reply:
Good point. Fixed in https://github.com/rust-osdev/x86_64/pull/258/commits/c47d482df86ffe43e7ac9fccc6716f2b74538849. I used "at the end of our function" instead of `1:` (as that label means nothing in a doc comment).